### PR TITLE
SSL Multi-loss Update

### PR DESCRIPTION
**Pull request by sam1373**
# What does this PR do ?

- adds multi-loss training to ssl, with corresponding list of decoders and losses defined through the ssl model config
- adds option to use pseudo-labels from quantization module of contrastive loss as targets for other losses
- adds option for ssl decoders to use outputs of non-final layer of encoder as input
- default ssl config is updated to use contrastive + mlm loss combination
- adds tests for ssl and maskedpatchaugmentation
- fixes sampling for tarred datasets with multiple utterances per file

**Collection**: ASR

**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

### CODE REVIEW

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall looks great, minor suggestions here and there. One or two major things to look at - regarding the offset stuff in the tarred data loader

File: nemo/collections/common/parts/preprocessing/collections.py [hunk 1]
```diff
@@ -202,7 +208,17 @@ def __init__(self, manifests_files: Union[str, List[str]], *args, **kwargs):
             **kwargs: Kwargs to pass to `AudioText` constructor.
         """
 
-        ids, audio_files, durations, texts, offsets, speakers, orig_srs, langs = [], [], [], [], [], [], [], []
+        ids, audio_files, durations, texts, offsets, speakers, orig_srs, token_labels, langs = (
```
[the remark this answers is not in the document]
Comment (sam1373) [reply nemo/collections/common/parts/preprocessing/collections.py:211]: This seems to be caused by style fix

File: nemo/collections/asr/data/audio_to_text.py [hunk 2]
```diff
@@ -142,7 +142,7 @@ def process_text_by_id(self, index: int) -> (List[int], int):
         return self.process_text_by_sample(sample)
 
     def process_text_by_file_id(self, file_id: str) -> (List[int], int):
-        manifest_idx = self.collection.mapping[file_id]
+        manifest_idx = self.collection.mapping[file_id][0]
```
[the remark this answers is not in the document]
Comment (sam1373) [reply nemo/collections/asr/data/audio_to_text.py:145]: Before this PR this was always a single id. Now mapping is a list, so this change just keeps previous functionality for dali

File: examples/asr/conf/ssl/conformer/conformer_ssl.yaml [hunk 3]
```diff
@@ -114,24 +114,38 @@ model:
...
+        # (quantizer is required to extract pseudo-labels for other losses)
+        codebook_size: 300
+        sample_from_same_utterance_only: true #should negatives be sampled only from the same utterance
+        sample_from_non_masked: false #should negatives be sampled from non-masked steps
+
+    mlm:
+      decoder:
+        _target_: nemo.collections.asr.modules.ConvASRDecoder
+        feat_in: ${model.encoder.d_model}
+        num_classes: 90000
```
[the remark this answers is not in the document]
Comment (sam1373) [reply examples/asr/conf/ssl/conformer/conformer_ssl.yaml:145]: This can be inferred as codebook_size^groups, but I don't think we can use formulas in the config so I just added a comment for now

File: nemo/collections/asr/data/audio_to_text.py [hunk 4]
```diff
@@ -675,17 +676,45 @@ def __next__(self):
...
     def _build_sample(self, tup):
         """Builds the training sample by combining the data from the WebDataset with the manifest info.
         """
-        audio_bytes, audio_filename = tup
+        audio_bytes, audio_filename, offset_id = tup
 
         # Grab manifest entry from self.manifest_preprocessor.collection
         file_id, _ = os.path.splitext(os.path.basename(audio_filename))
-        manifest_idx = self.manifest_processor.collection.mapping[file_id]
+        manifest_idx = self.manifest_processor.collection.mapping[file_id][offset_id]
```
[the remark this answers is not in the document]
Comment (sam1373) [reply nemo/collections/asr/data/audio_to_text.py:720]: It should always be there now.

File: nemo/collections/asr/data/audio_to_text.py [hunk 5]
```diff
@@ -675,17 +676,45 @@ def __next__(self):
 
         return TarredAudioFilter(self.manifest_processor.collection)
 
+    def _loop_offsets(self, iterator):
```
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/asr/data/audio_to_text.py:679]: I meant why it is needed, document that.

File: nemo/collections/asr/losses/ssl_losses/__init__.py [hunk 6]
```diff
@@ -12,4 +12,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from nemo.collections.asr.losses.pt_losses.contrastive import ContrastiveLoss
+from nemo.collections.asr.losses.ssl_losses.contrastive import ContrastiveLoss
```
Comment (titu1994) [inline nemo/collections/asr/losses/ssl_losses/__init__.py:15]: Will your pretrained model still work if this name is changed ?

File: nemo/collections/asr/losses/ssl_losses/contrastive.py [hunk 7]
```diff
@@ -142,7 +156,26 @@ def forward(self, spectrograms, spec_masks, decoder_outputs):
...
 
         if self.quantized_targets:
-            targets, prob_ppl_loss, cur_codebook_temp = self.quantizer(targets)
+            if self.store_ids:
+                # store ids for use by other losses
+                targets, prob_ppl_loss, cur_codebook_temp, self.target_ids = self.quantizer(targets, return_ids=True)
+
+                if self.reduce_ids:
+                    # reduce consecutive equivalent ids to a single occurence
+                    unique_x, indices = torch.unique_consecutive(self.target_ids, return_inverse=True)
```
Comment (titu1994) [inline nemo/collections/asr/losses/ssl_losses/contrastive.py:165]: Wow, cool function. Didn't know it existed.

File: nemo/collections/asr/losses/ssl_losses/contrastive.py [hunk 8]
```diff
@@ -238,6 +271,8 @@ def forward(self, spectrograms, spec_masks, decoder_outputs):
         if not isinstance(loss, torch.Tensor):
             loss = torch.Tensor([0]).to(device=decoder_outputs.device)
 
+        loss *= self.multiplier / spectrograms.shape[0]
```
Comment (titu1994) [inline nemo/collections/asr/losses/ssl_losses/contrastive.py:274]: Isn't shape(0) just batch size? Write that explicitly v

File: nemo/collections/asr/models/ssl_models.py [hunk 9]
```diff
@@ -278,38 +326,141 @@ def forward(
...
             spec_masks[idx, :, proc_len:] = 0.0
 
         encoded, encoded_len = self.encoder(audio_signal=processed_signal, length=processed_signal_length)
-        outputs = self.decoder_ssl(encoder_output=encoded)
 
-        return spectrograms, spec_masks, outputs
+        return spectrograms, spec_masks, encoded, encoded_len
+
+    def decoder_loss_step(self, spectrograms, spec_masks, encoded, encoded_len, targets=None, target_lengths=None):
+        loss_val_dict = {}
```
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/asr/models/ssl_models.py:339]: When I say documentation - I mean the high level info of what the function is doing - it's input arguments in Google style formatting (see other docstrings in ASR), it's output returns if anything and any additional notes of value. Please write full fledged docs.

File: nemo/collections/common/parts/preprocessing/collections.py [hunk 10]
```diff
@@ -150,28 +151,33 @@ def __init__(
...
-                    text_tokens = parser(text)
-            else:
-                text_tokens = []
+                    text_tokens = []
 
-            if text_tokens is None:
-                duration_filtered += duration
-                num_filtered += 1
-                continue
+                if text_tokens is None:
```
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/common/parts/preprocessing/collections.py:168]: Ok

File: nemo/collections/common/parts/preprocessing/collections.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/common/parts/preprocessing/collections.py:211]: Yes that's why I said instead of adding more items with comma to the above list, just create a new list below it.

**Review by titu1994 — COMMENTED** [review]
More comments. It looks much better.

File: nemo/collections/asr/losses/ssl_losses/__init__.py (hunk 6 above)
Comment (sam1373) [reply nemo/collections/asr/losses/ssl_losses/__init__.py:15]: Tested, seems to still work.

File: examples/asr/conf/ssl/contextnet/contextnet_ssl.yaml [hunk 11]
```diff
@@ -370,76 +370,93 @@ model:
...
   optim:
-    name: novograd
-    lr: 0.05
-
+    name: adamw
+    lr: 5.0
     # optimizer arguments
-    betas: [0.9, 0.0]
-    weight_decay: 0.001
+    betas: [ 0.9, 0.98 ]
```
Comment (titu1994) [inline examples/asr/conf/ssl/contextnet/contextnet_ssl.yaml:411]: Padding?

**Review by titu1994 — COMMENTED** [review]
Looks great ! Ready to merge

**Review by titu1994 — DISMISSED** [review]
